sherpa/models/tests/test_model.py

```diff
@@ -389,6 +393,86 @@ def test_get_mins_maxes_warns(caplog):
     assert msgs[5] == 'value of parameter box1d.xlow maximum is below hard minimum; setting to hard minimum'
 
 
+def test_set_hard_limit_warns(caplog):
+    """Setting the thawed pars is different to changing the par directly.
+
+    Check one of these cases. Created to address #1688.  Unfortunately
+    the check using caplog captures the logging output correctly, and
+    not as a used would see it (where the %s was not being replaced),
+    which means this can not check #1688 is fixed. DJB does not know
+    how to do such a check witout adding extensive testing machinery
+    which is not warranted in this case.
+
+    """
+
+    mdl = Sin("m")
+
+    # Check the defaults haven't changed.
+    #
+    assert mdl.thawedpars == pytest.approx([1, 0, 1])
+
+    # Check minimum values.
+    #
+    with caplog.at_level(logging.INFO, logger='sherpa'):
+        mdl.thawedpars = [0, -1, -1]
+
+    assert mdl.thawedpars == pytest.approx([1e-10, 0, 1e-5])
+    assert len(caplog.records) == 3
+
+    # Check maximum values.
+    #
+    big = 4e40
+    with caplog.at_level(logging.INFO, logger='sherpa'):
+        mdl.thawedpars = [big, big, big]
+
+    assert mdl.thawedpars == pytest.approx([10, hugeval, hugeval])
+    assert len(caplog.records) == 6
+
+    msgs = []
+    for (log_name, log_level, message) in caplog.record_tuples:
+        assert log_level == logging.WARNING
+        assert log_name == 'sherpa.models.model'
+        msgs.append(message)
+
+    def minstr(name):
+        return f"value of parameter {name} is below minimum; setting to minimum"
+
+    def maxstr(name):
+        return f"value of parameter {name} is above maximum; setting to maximum"
+
+    assert msgs[0] == minstr("m.period")
+    assert msgs[1] == minstr("m.offset")
+    assert msgs[2] == minstr("m.ampl")
+    assert msgs[3] == maxstr("m.period")
+    assert msgs[4] == maxstr("m.offset")
+    assert msgs[5] == maxstr("m.ampl")
+
+
+def test_set_pars_errors_soft_limits():
```

Comment:
I'm not quite sure what you want to test here: `test_set_pars_errors_soft_limits` is basically

```python
mdl = Box1D("b")
mdl.ampl.set(0.5, min=0, max=1)

# check we error our when try to set mdl.ampl < soft min
# check mdl.ampl.val is still 0.5
# check we error out when try to set mdl.ampl > soft.max
# check mdl.ampl.val is still 0.5
```
 
Are you thinking of adding a test of just

```
mdl.ampl = some-value
# have newval, minval, maxval either all > or < some-value
mdl.ampl.set(newval, min=minval, max=maxval)
```

which is a very different test (and possibly something to look at as part of #1700 but could be added here). Or am I missing something completely obvious?

Reply:
I'm thinking of
```
mdl = Box1D("b")
mdl.ampl.set(0.5, min=1e-6, max=1)

# check we error our when try to set hard_min < mdl.ampl < soft min
mdl.ampl = 1e-7
# check mdl.ampl.val is still 0.5
```
As written, we don't know if this test tests setting mdl.ampl < soft min or mdl.ampl < hard min, because in this situation hard min and soft min are identical. (The hard min is 0). So, instead of setting the `min=0`, set `min=1e-6` so that it's different from the hard min. Then, we can try setting to 1e-7 and we can be sure that any error warning truly results from setting below the soft min (as the name of the tests indicate) and not from setting below the hardmin. As written now, we don't know where the error comes from. Since the hardlin is also 0, the result you check for could also come from setting below the hard lim.

Reply:
Sorry, you had confused with your ordering. Note that the hard min/maxes are very different from the soft limits shown here - it depends on the parameter but for the ampl parameter here it's ~ +/- 3e-38

```
>>> from sherpa.models.basic import Box1D
>>> mdl = Box1D("m")
>>> print(mdl.ampl)
val         = 1.0
min         = -1.0
max         = 1.0
units       = 
frozen      = False
link        = None
default_val = 1.0
default_min = -1.0
default_max = 1.0
>>> mdl.ampl.hard_min
-3.4028234663852886e+38
>>> mdl.ampl.hard_max
3.4028234663852886e+38
```

Reply:
Mmm. I thought I had interactively loaded the model and checked. Maybe I messed that up.

Reply:
Maybe we should look at this as part of #1700 ?

Reply:
Sure.